Change disruption dynamic to not be stuuuuuupid

Ferociousfeind opened the pull request:
**Bugfix:** I suppose this is the ES equivalent to controversial politics, but I am declaring this behavior a bug, and I want to fix it.

## Fix Details
"Why do shields block shield disruption? Now they don't! This will make shield disruptions behave slightly differently."

EDIT: (Perhaps I should be more on top of updating the OPs of my PRs) The "ScaleType" function is modified to take two arguments- a fraction of damage deflected by shields, a fraction of damage deflected by hull, and the relevant damage-protection value. When shields are partly disrupted (or the projectile breaks shields and so some hits shields and some hits hull) the amount of damage deflected is interpolated linearly between the shield value and hull value.

## Testing Done
I am on vacation at the moment, with no means to test. I am fairly sure this is the correct implementation. The desired effect is for negative "blocking" to cause hull to block damage and for shields to take full damage. There is the possibility that this new behavior may require shield disruptors to be re-balanced, because they gained effectiveness as their target became more disrupted (more hull to hit, more disruption landed) whereas now they will gradually lose effectiveness as their target becomes more disrupted (more shields get bypassed, less disruption damage actually lands in the shields)

Until I get home to test, or someone else harasses me with their own tests, this will stay a draft.

Zitchas commented:
I have always just assumed that disruption was applying against the systems that make shields.

Ideally, *every* weapon should have an attribute that specifies how well this effect works against shields and a similar value for effectiveness against hull. So weapon A might do disruption best against shields, while weapon B did it best against hull. And have similar pairs for heat and every other social effect. Slowing, ionisation, etc

Amazinite commented on source/DamageProfile.cpp:
```diff
@@ -115,7 +115,10 @@ void DamageProfile::PopulateDamage(DamageDealt &damage, const Ship &ship) const
 	// the value of its protection attribute.
 	auto ScaleType = [&](double blocked, double protection)
```
Instead of having negative values somehow flip into using a different behavior, I'd prefer if this instead accepted a shield blocked and a hull blocked variable. If the shields are still up, use the shield blocked. If the shields are down, use the hull blocked.

If this change is being made to disruption, it should perhaps also be made to discharge. Discharge already currently deals 100% of its damage while the shields are up.

Ferociousfeind replied:
Okay, I've rewritten it to take inputs "shieldBlocked", "hullBlocked", and of course "protection", which based on "shieldFraction" will return the amount you'd expect. I could make discharge damage be 50% blocked by hull (because then there's no more shields to discharge, just as disruption is halved when there are no shields to disrupt).

I originally wrote it with a secondary behavior with a negative input because I didn't see much use in, say, shieldBlocked = 50% and hullBlocked _also_ = 50%. Just go through the data files and halve that damage type across the board. But this is more clear.

Ferociousfeind commented:
Oh gosh, hold on. Not all calls to "ScaleType" were properly updated. Hold on.

Amazinite commented:
You may need to update tooltips for these damage types.

Ferociousfeind commented:
Fixed a merge conflict and added a blurb to shield disruption/sec and shield disruption/shot (and fixed a small typo I spotted on the way, a "disruption" instead of a "discharge")

Where the tooltip did not mention that disruption fades over time, I did not change the tooltip.

Amazinite commented on source/DamageProfile.cpp:
```diff
@@ -111,11 +111,14 @@ void DamageProfile::PopulateDamage(DamageDealt &damage, const Ship &ship) const
+	auto ScaleType = [&](double shieldBlocked, double hullBlocked, double protection)
 	{
-		return damage.scaling * (1. - blocked * shieldFraction) / (1. + protection);
+		return damage.scaling *
+			(((1. - shieldBlocked) * (1. - shieldFraction))
+			+ ((1. - hullBlocked) * (shieldFraction)))
+			/ (1. + protection);
```
As best I can tell, this equation does not work at all and should look something more like this.
```suggestion
		double blocked = min(1 - shieldBlocked * shieldFraction, 1 - hullBlocked * (1 - shieldFraction));
		return damage.scaling * blocked / (1. + protection);
```
At least in the cases where shield fraction is 1 or 0 (i.e. your shields entirely block shield damage or your shields are entirely down), this works. We could definitely put extra scrutiny on whether this works for shield fraction values between 0 and 1, though, but from my light messing around in Desmos, this seems to work.

Ferociousfeind replied:
This equation mirrors what was already in place, with an implicit "hull never blocks anything" assumption. If a damage type is 50% blocked by shields and 25% blocked by hull, and the target's shields are 50% disrupted, then the damage type should be blocked 37.5%- 1/2 of 50% plus 1/2 of 25%.

What's in the PR now may be using shieldFraction backwards... If shieldFraction is equal to 1 when shields are up and there is no disruption (i.e. "the fraction of the projectile hitting shields" is 1), then my formula is wrong, and the uses of `shieldFraction` and `(1. - shieldFraction)` should be swapped.

Your proposal currently behaves opposite mine when dumped into desmos and only one blockage value is nonzero (which is why I suspect I have `shieldFraction` backwards), but when both are nonzero, the behavior is strange. It's, of course, the minimum of two lines, one pinned to (0, 1), the other pinned to (1, 1) (where, uh, X is shield fraction, Y is % of damage that gets through shields). As an extreme example, if a damage type is said to be 100% blocked by shields and also 100% blocked by hull (useful, I know), your proposal would have a half-disrupted shield allow 50% of the damage to leak through, instead of 0% damage. Because the hull is only half visible, and would allow 50% damage, and the shields are half-visible and would allow 50% damage. Min between 50 and 50 is 50, rather than 0. The use of `min` is _very_ strange to me...

Anyways, I will swap the uses of `shieldFraction` and `(1. - shieldFraction)`, and that should fix the strange behavior you're mentioning.

Amazinite replied:
Even when I swapped the shield fraction to what it should have been in your question, it didn't look correct to me. I think it'd be far easier if instead of trying to cram everything into a single equation, we just have a "shield fraction == 1" case (`1 - shieldBlocked`), "shield fraction == 0" case (`1 - hullBlocked`), then a case for when shield fraction is something in between.

Amazinite commented on source/DamageProfile.cpp:
```diff
@@ -111,11 +111,14 @@ void DamageProfile::PopulateDamage(DamageDealt &damage, const Ship &ship) const
+	auto ScaleType = [&](double shieldBlocked, double hullBlocked, double protection)
 	{
-		return damage.scaling * (1. - blocked * shieldFraction) / (1. + protection);
+		return damage.scaling *
+			(((1. - shieldBlocked) * (shieldFraction))
+			+ ((1. - hullBlocked) * (1. - shieldFraction)))
+			/ (1. + protection);
```
Played around with this a bit more after the fix, and I think it checks out now. I do think we should bust it out into its own line for readability, though.
```suggestion
		double blocked = (1. - shieldBlocked) * (shieldFraction) + (1. - hullBlocked) * (1. - shieldFraction);
		return damage.scaling * blocked / (1. + protection);
```
